handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -822,11 +843,31 @@ private void wrap(ChannelHandlerContext ctx, boolean inUnwrap) throws SSLExcepti
                     break;
                 }
 
-                if (out == null) {
-                    out = allocateOutNetBuf(ctx, buf.readableBytes(), buf.nioBufferCount());
+                SSLEngineResult result;
+
+                if (buf.readableBytes() > MAX_PLAINTEXT_LENGTH) {
+                    // If we pulled a buffer larger than the supported packet size, we can slice it up and iteratively,
+                    // encrypting multiple packets into a single larger buffer. This substantially saves on allocations
+                    // for large responses. Here we estimate how large of a buffer we need. If we overestimate a bit,
+                    // that's fine. If we underestimate, we'll simply re-enqueue the remaining buffer and get it on the
+                    // next outer loop.
+                    int readableBytes = buf.readableBytes();
+                    int numPackets = readableBytes / MAX_PLAINTEXT_LENGTH;
+                    if (readableBytes % MAX_PLAINTEXT_LENGTH != 0) {
+                        numPackets += 1;
+                    }
+
+                    if (out == null) {
+                        out = allocateOutNetBuf(ctx, readableBytes, buf.nioBufferCount() + numPackets);
```

Comment:
In `wrapMultiple`, we will pass slices up to `MAX_PLAINTEXT_LENGTH` in length to `engine.wrap()`. So we will create at least `numPackets` encrypted packets. If that were it, we could just pass numPackets here, and `readableBytes + numPackets * wrapOverhead` would be an accurate measure of the total output buffer needed.

But also, `engine.wrap()` takes a `ByteBuffer[]` and wraps each element separately. Your input buffer's backing `nioBuffers()` may not align perfectly to `MAX_PLAINTEXT_LENGTH`. So there might be a case where we pass 16k to engine.wrap() but across 2+ buffers, thus 2+ packets for what we thought was one.

So this attempts to account for both of those constraints, albeit very imperfectly. I'm open to other options here, but I didn't try too hard because the estimate doesn't need to be perfect. If we underestimate, we'll just end up re-enqueing the remaining bytes and picking them up on the next outer while loop of wrap. If we overestimate, there might be some extra unused bytes at the end of our OutNetBuf which will be reclaimed once we finish.

Reply:
I see, hence the comments about over-/under-estimating.

Reply:
I could see a case being made for deciding to just use numPackets, which would likely often be an underestimate whenever input buffers have more than 1 nioBuffer. The downside of underestimating is we need to re-enqueue the remaining portion of the input buffer back into SslHandlerCoalescingBufferQueue and re-poll it, allocate a new smaller buffer, etc. If this happens once, it's not a big deal. If it happens many times, we start to regress back to the original design. The downside of over estimation is we waste a bit of memory temporarily until it can be released by the AbstractUnsafe.flush0.  That would be bad if we could over estimate by a lot, but I don't think we will in most cases -- we might be off by a few packets * 100.